BridgeSDK/BridgeAPI/SBBActivityManager.m

```diff
@@ -65,75 +67,99 @@ - (NSURLSessionDataTask *)getScheduledActivitiesForDaysAhead:(NSInteger)daysAhea
                 if (completion) {
-                    completion(tasks, nil);
+                    NSMutableArray *requestedTasks = [[self filterTasks:tasks.items forDaysAhead:daysAhead andDaysBehind:daysBehind excludeStillValid:NO] mutableCopy];
+                    if (isInternalObjectManager) {
+                        // in case object mapping has been set up
+                        for (NSInteger i = 0; i < requestedTasks.count; ++i) {
+                            requestedTasks[i] = [((id<SBBObjectManagerInternalProtocol>)self.objectManager) mappedObjectForBridgeObject:requestedTasks[i]];
+                        }
+                    }
+                    completion(requestedTasks, nil);
                 }
             });
             return nil;
         }
         
-        // otherwise, keep unfinished tasks from yesterday so we can show them in the Yesterday section
+        // otherwise, keep all tasks from the past kDaysToCache days, and finished ones from today (or later?)
         // (because the server doesn't give us those)
-        if (daysBehind > 0) {
-            NSCalendar *cal = [NSCalendar currentCalendar];
-            NSDate *today = [cal startOfDayForDate:[NSDate date]];
-            NSDate *windowStartDate = [cal startOfDayForDate:[NSDate dateWithTimeIntervalSinceNow:-daysBehind * kSBB24Hours]];
-            NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(id  _Nonnull evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
-                SBBScheduledActivity *activity = (SBBScheduledActivity *)evaluatedObject;
-                if (!activity.expiresOn) {
-                    // if it never expires it would still be in the list from the server
-                    return NO;
-                }
-                
-                NSComparisonResult compareToToday = [activity.expiresOn compare:today];
-                NSComparisonResult compareToStart = [activity.expiresOn compare:windowStartDate];
-                BOOL withinDaysBehind = ((compareToToday == NSOrderedAscending) || (compareToToday == NSOrderedSame)) && ((compareToStart == NSOrderedDescending) || (compareToStart == NSOrderedSame));
-                
-                return withinDaysBehind && activity.finishedOn == nil;
-            }];
-            
-            daysBehindTasks = [tasks.items filteredArrayUsingPredicate:predicate];
-        }
+        savedTasks = [self filterTasks:tasks.items forDaysAhead:kMaxAdvance andDaysBehind:kDaysToCache excludeStillValid:YES];
     }
 
     return [self.networkManager get:kSBBActivityAPI headers:headers parameters:@{@"daysAhead": @(daysAhead), @"offset": [[NSDate date] ISO8601OffsetString]} completion:^(NSURLSessionDataTask *task, id responseObject, NSError *error) {
-        tasks = [self.objectManager objectFromBridgeJSON:responseObject];
+        SBBResourceList *tasks = [self.objectManager objectFromBridgeJSON:responseObject];
```

Comment:
The objectFromBridgeJSON call is how objects get cached or have their cached representations updated; that's the reason I call it here. Although always cached as Bridge objects, what that method returns may potentially have been mapped to a different object type, if mappings have been set up in this instance of the object manager. So below, once I know the cached version is current, I pull it directly from the cache to make sure I always am working with genuine SBB objects while I'm processing everything. Then below I call the mapping routine to make sure I'm passing the expected objects to the completion handler.

Since no one is actually using the mapping feature AFAIK, it might be worth considering just pulling it out to simplify maintenance going forward, but that would be a job for its own separate pull request.